### PR TITLE
DDF-3101 changed regex to check for hostname only connection to STS, added test for constraint interceptor

### PR DESCRIPTION
**Pull request by stustison**
#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged)?
@ahoffer @lessarderic @brendan-hofmann 
#### Select relevant component teams: 
@codice/security 
#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@clockard
@stustison
#### What are the relevant tickets?
[](https://codice.atlassian.net/browse/)
#### Checklist:
- [ ] Documentation Updated
- [ ] Change Log Updated
- [x] Update / Add Unit Tests
- [ ] Update / Add Integration Tests
#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

**Comment by stustison** [comment]
build now

File: platform/security/sts/security-sts-server/src/test/java/ddf/security/sts/SubjectDNConstraintsInterceptorTest.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        when(message.get(WSHandlerConstants.SIG_SUBJECT_CERT_CONSTRAINTS)).thenReturn("CN=cert");
+        Request request = mock(Request.class);
+        when(message.get("HTTP.REQUEST")).thenReturn(request);
+        X509Certificate x509Certificate = mock(X509Certificate.class);
+        X500Principal x500Principal = new X500Principal("CN=cert");
+        when(x509Certificate.getSubjectX500Principal()).thenReturn(x500Principal);
+        when(request.getAttribute("javax.servlet.request.X509Certificate")).thenReturn(new X509Certificate[] {
+                x509Certificate});
+
+        subjectDNConstraintsInterceptor.handleMessage(message);
```
Comment (brendan-hofmann) [inline platform/security/sts/security-sts-server/src/test/java/ddf/security/sts/SubjectDNConstraintsInterceptorTest.java]: ✏️ The javadoc says it returns true on success, but the signature is void. If the signature is correct, I guess there's nothing to really assert here, but maybe we should update the javadoc?

Comment (stustison) [inline platform/security/sts/security-sts-server/src/test/java/ddf/security/sts/SubjectDNConstraintsInterceptorTest.java]: The signature is correct.

**Comment by stustison** [comment]
build now

File: distribution/ddf-common/src/main/resources/etc/system.properties [hunk 2]
```diff
@@ -284,7 +284,8 @@ karaf.secured.services=(&(osgi.command.scope=*)(osgi.command.function=*))
 #java.security.policy=${karaf.home}/etc/all.policy
 #org.osgi.framework.security=osgi
 #org.osgi.framework.trust.repositories=${karaf.home}/etc/trustStore.ks
-ws-security.subject.cert.constraints = .*
+#ws-security.subject.cert.constraints = pipe (|) separated list of regex patterns for DNs that are allowed to connect to the Security Token Service
```
Comment (lessarderic) [inline distribution/ddf-common/src/main/resources/etc/system.properties]: ✏️  Nit: To be consistent with the other comments in the file, should we aviod putting the property name in the comment and put a space between the `#` and the description?

File: platform/security/sts/security-sts-server/src/test/java/ddf/security/sts/SubjectDNConstraintsInterceptorTest.java (hunk 1 above)
Comment (lessarderic) [inline platform/security/sts/security-sts-server/src/test/java/ddf/security/sts/SubjectDNConstraintsInterceptorTest.java]: ❗️ We should update the `SubjectDNConstraintsInterceptor.handleMessage()` comment.

File: platform/security/sts/security-sts-server/src/test/java/ddf/security/sts/SubjectDNConstraintsInterceptorTest.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+    SubjectDNConstraintsInterceptor subjectDNConstraintsInterceptor;
+
+    @Before
+    public void setup() {
+        subjectDNConstraintsInterceptor = new SubjectDNConstraintsInterceptor();
+    }
+
+    @Test
+    public void testHandleMessage() {
```
Comment (lessarderic) [inline platform/security/sts/security-sts-server/src/test/java/ddf/security/sts/SubjectDNConstraintsInterceptorTest.java]: ✏️  Should we add `FromAllowedDn` to the name to clarify what this is testing?

File: platform/security/sts/security-sts-server/src/test/java/ddf/security/sts/SubjectDNConstraintsInterceptorTest.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        X500Principal x500Principal = new X500Principal("CN=cert");
+        when(x509Certificate.getSubjectX500Principal()).thenReturn(x500Principal);
+        when(request.getAttribute("javax.servlet.request.X509Certificate")).thenReturn(new X509Certificate[] {
+                x509Certificate});
+
+        subjectDNConstraintsInterceptor.handleMessage(message);
+    }
+
+    @Test
+    public void testHandleMessageCommaPattern() {
```
Comment (lessarderic) [inline platform/security/sts/security-sts-server/src/test/java/ddf/security/sts/SubjectDNConstraintsInterceptorTest.java]: ✏️  Rename to `testHandleMessageWithMultipleCommaPatternsFromAllowedDN` or something along those lines?

File: platform/security/sts/security-sts-server/src/main/java/ddf/security/sts/SubjectDNConstraintsInterceptor.java [hunk 5]
```diff
@@ -74,8 +75,8 @@ public void handleMessage(Message message) throws Fault {
 
     public Collection<Pattern> setSubjectDNPatterns(String subjectDNConstraints) {
```
Comment (lessarderic) [inline platform/security/sts/security-sts-server/src/main/java/ddf/security/sts/SubjectDNConstraintsInterceptor.java]: ❗️ Is there a reason for this to be `public`? Couldn't find any other class calling this method.

Comment (lessarderic) [inline platform/security/sts/security-sts-server/src/main/java/ddf/security/sts/SubjectDNConstraintsInterceptor.java]: ✏️  The name of this method is somewhat confusing. I was expecting it to set something on the object but it really just converts the `subjectDNConstraints` and returns a list of `Pattern`. Should we consider renaming it?

File: platform/security/sts/security-sts-server/src/main/java/ddf/security/sts/SubjectDNConstraintsInterceptor.java [hunk 6]
```diff
@@ -74,8 +75,8 @@ public void handleMessage(Message message) throws Fault {
...
         ArrayList<Pattern> subjectDNPatterns = new ArrayList<>();
-        if (subjectDNConstraints != null) {
-            String[] patterns = subjectDNConstraints.split(",");
+        if (StringUtils.isNotEmpty(subjectDNConstraints)) {
```
Comment (lessarderic) [inline platform/security/sts/security-sts-server/src/main/java/ddf/security/sts/SubjectDNConstraintsInterceptor.java]: ❓ Assuming we make this method `private`, can we replace the check on L56 with this one and get rid of this `if` statement? Or only have the check here, return an empty `List` if the constraint is `null` or empty and have the calling code trigger on that instead of the `subjectDNContraints` value? Just trying to avoid double `subjectDNContraints` validations that could become out-of-sync in the future.

File: platform/security/sts/security-sts-server/src/test/java/ddf/security/sts/SubjectDNConstraintsInterceptorTest.java [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        when(request.getAttribute("javax.servlet.request.X509Certificate")).thenReturn(new X509Certificate[] {
+                x509Certificate});
+
+        subjectDNConstraintsInterceptor.handleMessage(message);
+    }
+
+    @Test
+    public void testHandleMessageNoPattern() {
+        Message message = mock(Message.class);
+        when(message.get(WSHandlerConstants.SIG_SUBJECT_CERT_CONSTRAINTS)).thenReturn(null);
```
Comment (lessarderic) [inline platform/security/sts/security-sts-server/src/test/java/ddf/security/sts/SubjectDNConstraintsInterceptorTest.java]: ❗️ We need to test the case where the constraints are empty.

File: platform/security/sts/security-sts-server/src/test/java/ddf/security/sts/SubjectDNConstraintsInterceptorTest.java [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+        when(message.get(WSHandlerConstants.SIG_SUBJECT_CERT_CONSTRAINTS)).thenReturn("CN=certbad");
+        Request request = mock(Request.class);
+        when(message.get("HTTP.REQUEST")).thenReturn(request);
+        when(request.getAttribute("javax.servlet.request.X509Certificate")).thenReturn(null);
+
+        subjectDNConstraintsInterceptor.handleMessage(message);
+    }
+
+    @Test
+    public void testSetSubjectDNPatterns() {
```
Comment (lessarderic) [inline platform/security/sts/security-sts-server/src/test/java/ddf/security/sts/SubjectDNConstraintsInterceptorTest.java]: ✏️  There's a lot of repeated setup code in this test class. We should consider extracting that code to a private method that takes parameters, use JUnit parametrized tests, or use Spock. Going with parametrized tests or Spock would help capture all the pattern tests in `testSetSubjectDNPatterns` (which really should be separate tests) and not have to expose the `setSubjectDNPatterns` method.

File: platform/security/sts/security-sts-server/src/main/java/ddf/security/sts/SubjectDNConstraintsInterceptor.java (hunk 5 above)
Comment (lessarderic) [inline platform/security/sts/security-sts-server/src/main/java/ddf/security/sts/SubjectDNConstraintsInterceptor.java]: ✏️  Per our coding standards, we should camel-case acronyms, so the class, variable, parameters and method names should `Dn` instead of `DN`. Not a big deal if it's too much trouble to rename all of those.

**Comment by stustison** [comment]
build now

**Comment by stustison** [comment]
build now

**Comment by stustison** [comment]
build now
